pkg/tsdb/elasticsearch/response_parser.go

```diff
@@ -56,19 +57,156 @@ func parseResponse(responses []*es.SearchResponse, targets []*Query) (*backend.Q
 
 		queryRes := backend.DataResponse{}
 
-		props := make(map[string]string)
-		err := processBuckets(res.Aggregations, target, &queryRes, props, 0)
-		if err != nil {
-			return &backend.QueryDataResponse{}, err
-		}
-		nameFields(queryRes, target)
-		trimDatapoints(queryRes, target)
+		if isDocumentQuery(target) {
+			err := processDocumentResponse(res, target, timeField, &queryRes)
+			if err != nil {
+				return &backend.QueryDataResponse{}, err
+			}
+			result.Responses[target.RefID] = queryRes
+		} else {
+			props := make(map[string]string)
+			err := processBuckets(res.Aggregations, target, &queryRes, props, 0)
+			if err != nil {
+				return &backend.QueryDataResponse{}, err
+			}
+			nameFields(queryRes, target)
+			trimDatapoints(queryRes, target)
 
-		result.Responses[target.RefID] = queryRes
+			result.Responses[target.RefID] = queryRes
+		}
 	}
 	return &result, nil
 }
 
+func processDocumentResponse(res *es.SearchResponse, target *Query, timeField string, queryRes *backend.DataResponse) error {
+	docs := make([]map[string]interface{}, len(res.Hits.Hits))
+	propNames := make([]string, 0)
+
+	for hitIdx, hit := range res.Hits.Hits {
+		var flattened map[string]interface{}
+		if hit["_source"] != nil {
+			flattened = flatten(hit["_source"].(map[string]interface{}))
+		}
+
+		doc := map[string]interface{}{
+			"_id":       hit["_id"],
+			"_type":     hit["_type"],
+			"_index":    hit["_index"],
+			"sort":      hit["sort"],
+			"highlight": hit["highlight"],
+			"_source":   flattened,
+		}
+
+		for k, v := range flattened {
+			doc[k] = v
+		}
+
+		for key := range doc {
+			// We want to add only unique propNames
+			if isStringInSlice(key, propNames) {
+				continue
+			} else {
+				propNames = append(propNames, key)
+			}
+		}
+
+		docs[hitIdx] = doc
+	}
+
+	size := len(docs)
+	isFilterable := true
+	allFields := make([]*data.Field, len(propNames))
+
+	sortedPropNames := sortPropNames(propNames, timeField)
+
+	for propNameIdx, propName := range sortedPropNames {
+		// Special handling for time field
+		if propName == timeField {
+			timeVector := make([]time.Time, size)
+			for i, doc := range docs {
+				timeString := doc[timeField].(string)
```

Comment:
this to-string-conversion can panic if the value is not really a string.
i'd recommend to check the error and just `continue` if that's the case, something like:
```suggestion
				timeString, ok := doc[timeField].(string)
				if !ok {
				    continue
				}
```